## Desription
The nonprofit foundation Alphabet Soup wants a tool that can help it select the applicants for funding with the best chance of success in their ventures. With knowledge of machine learning and neural networks, I’ll use the features in the provided dataset to create a binary classifier that can predict whether applicants will be successful if funded by Alphabet Soup.

From Alphabet Soup’s business team, you have received a CSV containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. Within this dataset are a number of columns that capture metadata about each organization, such as:

* **EIN** and **NAME**—Identification columns
* **APPLICATION_TYPE**—Alphabet Soup application type
* **AFFILIATION**—Affiliated sector of industry
* **CLASSIFICATION**—Government organization classification
* **USE_CASE**—Use case for funding
* **ORGANIZATION**—Organization type
* **STATUS**—Active status
* **INCOME_AMT**—Income classification
* **SPECIAL_CONSIDERATIONS**—Special consideration for application
* **ASK_AMT**—Funding amount requested
* **IS_SUCCESSFUL**—Was the money used effectively

### Step 3: Optimize the Model

Using your knowledge of TensorFlow, optimize your model to achieve a target predictive accuracy higher than 75%.

Using any or all of the following methods to optimize your model:

* Adjust the input data to ensure that no variables or outliers are causing confusion in the model, such as:
  * Dropping more or fewer columns.
  * Creating more bins for rare occurrences in columns.
  * Increasing or decreasing the number of values for each bin.
* Add more neurons to a hidden layer.
* Add more hidden layers.
* Use different activation functions for the hidden layers.
* Add or reduce the number of epochs to the training regimen.

**Note**: If you make at least three attempts at optimizing your model, you will not lose points if your model does not achieve target performance.

1. Create a new Jupyter Notebook file and name it `AlphabetSoupCharity_Optimzation.ipynb`.

2. Import your dependencies and read in the `charity_data.csv` to a Pandas DataFrame.

3. Preprocess the dataset like you did in Step 1, Be sure to adjust for any modifications that came out of optimizing the model.

4. Design a neural network model, and be sure to adjust for modifications that will optimize the model to achieve higher than 75% accuracy.

5. Save and export your results to an HDF5 file. Name the file `AlphabetSoupCharity_Optimization.h5`.

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME' 'STATUS', 'SPECIAL_CONSIDERATIONS'.
application_df = application_df.drop(columns = ['EIN', 'NAME', 'STATUS', 'SPECIAL_CONSIDERATIONS'])
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION     INCOME_AMT   ASK_AMT  IS_SUCCESSFUL  
0       Association              0      5000              1  
1      Co-operative         1-9999    108590              1  
2       Association              0      5000              0  
3             Trust    10000-24999      6692              1  
4             Trust  100000-499999    142590              1  
...             ...            ...       ...            ...  
34294   Association              0      5000              0  
34295   Association              0      5000              0  
34296   Association              0      5000              0  
34297   Association              0      5000              1  
34298  Co-operative          1M-5M  36500179              0  

[34299 rows x 8 columns]

In [3]:
# Determine the number of unique values in each column
application_df.nunique()

APPLICATION_TYPE      17
AFFILIATION            6
CLASSIFICATION        71
USE_CASE               5
ORGANIZATION           4
INCOME_AMT             9
ASK_AMT             8747
IS_SUCCESSFUL          2
dtype: int64

In [4]:
# Drop the non-beneficial ID columns, 'ASK_AMT	'.
application_df = application_df.drop(columns = ['ASK_AMT'])
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION     INCOME_AMT  IS_SUCCESSFUL  
0       Association              0              1  
1      Co-operative         1-9999              1  
2       Association              0              0  
3             Trust    10000-24999              1  
4             Trust  100000-499999              1  
...             ...            ...            ...  
34294   Association              0              0  
34295   Association              0              0  
34296   Association              0              0  
34297   Association              0              1  
34298  Co-operative          1M-5M              0  

[34299 rows x 7 columns]

In [5]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE    17
AFFILIATION          6
CLASSIFICATION      71
USE_CASE             5
ORGANIZATION         4
INCOME_AMT           9
IS_SUCCESSFUL        2
dtype: int64

In [6]:
# Look at APPLICATION_TYPE value counts for binning
type_counts = application_df['APPLICATION_TYPE'].value_counts()
type_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(type_counts.iloc[8:].index)
cutoff_value = 528

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other_app_type")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3                27037
T4                 1542
T6                 1216
T5                 1173
T19                1065
T8                  737
T7                  725
T10                 528
Other_app_type      276
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
classification_types = application_df['CLASSIFICATION'].value_counts()
classification_types

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list(classification_types.iloc[5:].index)
cutoff_value_classifications = 1883

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other_classification")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000                   17326
C2000                    6074
C1200                    4837
Other_classification     2261
C3000                    1918
C2100                    1883
Name: CLASSIFICATION, dtype: int64

In [10]:
#convert the categorical data in numeric using get_dummies()
dummies = pd.get_dummies(application_df)
dummies.head()

IS_SUCCESSFUL  APPLICATION_TYPE_Other_app_type  APPLICATION_TYPE_T10  \
0              1                                0                     1   
1              1                                0                     0   
2              0                                0                     0   
3              1                                0                     0   
4              1                                0                     0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                     0                    0                    0   
1                     0                    1                    0   
2                     0                    0                    0   
3                     0                    1                    0   
4                     0                    1                    0   

   APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  \
0                    0                    0                    0   
1                    0                    0                    0   
2                    1                    0                    0   
3                    0                    0                    0   
4                    0                    0                    0   

   APPLICATION_TYPE_T8  ...  ORGANIZATION_Trust  INCOME_AMT_0  \
0                    0  ...                   0             1   
1                    0  ...                   0             0   
2                    0  ...                   0             1   
3                    0  ...                   1             0   
4                    0  ...                   1             0   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  
0                0                  0  
1                0                  0  
2                0                  0  
3                0                  0  
4                0                  0  

[5 rows x 40 columns]

In [11]:
# Split our preprocessed data into our features and target arrays
y = dummies['IS_SUCCESSFUL'].values
X = dummies.drop(columns='IS_SUCCESSFUL').values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [12]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [13]:
# FIRST ATTEMPT
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features_total = len(X_train[0])
hidden_nodes_layer1 = 8
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim = input_features_total, activation = "relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation = "relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 320       
                                                                 
 dense_1 (Dense)             (None, 5)                 45        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 371 (1.45 KB)
Trainable params: 371 (1.45 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [14]:
# Compile the model (FIRST ATTEMPT)
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [15]:
# Train the model (FIRST ATTEMPT)
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6154 - accuracy: 0.6903
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5661 - accuracy: 0.7290
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5600 - accuracy: 0.7286
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5571 - accuracy: 0.7289
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5549 - accuracy: 0.7284
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5529 - accuracy: 0.7296
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5520 - accuracy: 0.7311
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5512 - accuracy: 0.7306
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5505 - accuracy: 0.7322
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5498 - accura

In [16]:
# Evaluate the model using the test data (FIRST ATTEMPT)
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5545 - accuracy: 0.7325 - 340ms/epoch - 1ms/step
Loss: 0.5544877648353577, Accuracy: 0.732478141784668


In [17]:
# SECOND ATTEMPT
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features_total = len(X_train[0])
hidden_nodes_layer1 = 10
hidden_nodes_layer2 = 10

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim = input_features_total, activation = "relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation = "relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 10)                400       
                                                                 
 dense_4 (Dense)             (None, 10)                110       
                                                                 
 dense_5 (Dense)             (None, 1)                 11        
                                                                 
Total params: 521 (2.04 KB)
Trainable params: 521 (2.04 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [18]:
# Compile the model (SECOND ATTEMPT)
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [19]:
# Train the model (SECOND ATTEMPT)
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 1ms/step - loss: 0.6023 - accuracy: 0.6939
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5638 - accuracy: 0.7238
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5578 - accuracy: 0.7280
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5561 - accuracy: 0.7266
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5541 - accuracy: 0.7285
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5529 - accuracy: 0.7289
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5525 - accuracy: 0.7293
Epoch 8/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5518 - accuracy: 0.7297
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5511 - accuracy: 0.7299
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5506 - accura

In [20]:
# Evaluate the model using the test data (SECOND ATTEMPT)
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5565 - accuracy: 0.7335 - 326ms/epoch - 1ms/step
Loss: 0.5564689636230469, Accuracy: 0.7335277199745178


In [21]:
# THIRD ATTEMPT
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features_total = len(X_train[0])
hidden_nodes_layer1 = 20
hidden_nodes_layer2 = 20

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim = input_features_total, activation = "tanh"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation = "sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="relu"))

# Check the structure of the model
nn.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 20)                800       
                                                                 
 dense_7 (Dense)             (None, 20)                420       
                                                                 
 dense_8 (Dense)             (None, 1)                 21        
                                                                 
Total params: 1241 (4.85 KB)
Trainable params: 1241 (4.85 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [22]:
# Compile the model (THIRD ATTEMPT)
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [23]:
# Train the model (THIRD ATTEMPT)
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 1ms/step - loss: 0.6228 - accuracy: 0.6832
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5728 - accuracy: 0.7215
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5690 - accuracy: 0.7262
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5685 - accuracy: 0.7266
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5624 - accuracy: 0.7271
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5596 - accuracy: 0.7277
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5547 - accuracy: 0.7282
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5548 - accuracy: 0.7289
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5765 - accuracy: 0.7147
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5585 - accura

In [24]:
# Evaluate the model using the test data (THIRD ATTEMPT)
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5742 - accuracy: 0.7326 - 329ms/epoch - 1ms/step
Loss: 0.574219822883606, Accuracy: 0.7325947284698486


In [28]:
# Export our model to HDF5 file
from google.colab import files

nn.save("AlphabetSoupCharity_Optimization.h5")
files.download("AlphabetSoupCharity_Optimization.h5")


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### Overview

This report evaluates three different iterations of a deep learning neural network model designed for Alphabet Soup. The purpose of this analysis is to refine the model's accuracy in predicting a specific target based on a set of input features. The models differ in their architecture in terms of the number of neurons and layers.

### Results

Data Preprocessing

- Target Variable(s): The target variable, likely a binary classification, is inferred from the output layer's single neuron across all models.

- Feature Variable(s): The models use various input features, as indicated by the different input parameters in the first dense layer of each model.

- Variables to Remove: Specific variables to be removed are not mentioned, but generally, irrelevant or non-contributory data should be excluded.

Model Iterations and Performance

1. First Attempt:

- Architecture: 8 neurons (first layer), 5 neurons (second layer).

- Performance: Loss: 0.5545, Accuracy: 73.25%.

- Observation: A moderate start in terms of performance with a simple architecture.

2. Second Attempt:


- Architecture: 10 neurons each in first and second layers.

- Performance: Loss: 0.5565, Accuracy: 73.35%.

- Observation: A slight improvement in accuracy, though the loss increased marginally.

3. Third Attempt:
    

- Architecture: 20 neurons each in first and second layers.

- Performance: Loss: 0.5742, Accuracy: 73.26%.

- Observation: Increasing the number of neurons did not significantly improve performance; in fact, loss increased.

Summary and Recommendations

The three model iterations show marginal differences in accuracy, hovering around 73%. The increasing complexity in model architecture does not yield a proportional increase in performance. This outcome suggests a potential overfitting issue or the need for a more refined approach in feature selection and model tuning.

### Recommendations:

To improve the neural network model for Alphabet Soup, I would focus on refining input features through feature engineering and experimenting with hyperparameter tuning, such as adjusting activation functions and learning rates, to optimize performance. Additionally, exploring alternative machine learning models like Random Forest or Gradient Boosting may provide more accurate predictions for the given dataset.